File: kartwheel/app/src/main/java/us/handstand/kartwheel/fragment/GameInfoFragment.java
package us.handstand.kartwheel.fragment;


import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.Loader;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import butterknife.ButterKnife;
import us.handstand.kartwheel.R;
import us.handstand.kartwheel.activity.TicketActivity;
import us.handstand.kartwheel.model.AndroidStorage;
import us.handstand.kartwheel.model.loader.GameInfoLoader;


public class GameInfoFragment extends Fragment implements TicketActivity.TicketFragment, LoaderManager.LoaderCallbacks {

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View fragmentView = inflater.inflate(R.layout.fragment_game_info, container, false);
        ButterKnife.bind(this, fragmentView);
        return fragmentView;
    }

    @Override
    public void onClick(View v) {

    }

    @Override
    public Loader onCreateLoader(int id, Bundle args) {
        return new GameInfoLoader(getActivity(), AndroidStorage.getString(AndroidStorage.TEAM_ID));
    }

    @Override
    public void onLoadFinished(Loader loader, Object data) {

    }

    @Override
    public void onLoaderReset(Loader loader) {

    }
}

File: kartwheel/app/src/main/kotlin/us/handstand/kartwheel/fragment/GameInfoFragment.kt
package us.handstand.kartwheel.fragment


import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v4.app.ShareCompat
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import us.handstand.kartwheel.R
import us.handstand.kartwheel.activity.TicketActivity
import us.handstand.kartwheel.controller.GameInfoController
import us.handstand.kartwheel.controller.TicketController.Companion.FORFEIT
import us.handstand.kartwheel.controller.TicketController.Companion.GAME_INFO
import us.handstand.kartwheel.layout.GameInfoPlayerView
import us.handstand.kartwheel.layout.ViewUtil
import us.handstand.kartwheel.model.Database
import us.handstand.kartwheel.model.Storage
import us.handstand.kartwheel.model.Ticket
import us.handstand.kartwheel.model.User

class GameInfoFragment : Fragment(), TicketActivity.TicketFragment, GameInfoController.Companion.GameInfoCompletionListener, GameInfoPlayerView.Companion.OnPlayerActionClickListener {
    private var playerOne: GameInfoPlayerView? = null
    private var playerTwo: GameInfoPlayerView? = null
    private var controller: GameInfoController? = null

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        // TODO: FAB Buttons
        val fragmentView = inflater!!.inflate(R.layout.fragment_game_info, container, false) as ViewGroup
        playerOne = ViewUtil.findView(fragmentView, R.id.playerOne)
        playerTwo = ViewUtil.findView(fragmentView, R.id.playerTwo)
        playerOne!!.setOnForfeitClickListener(this)
        playerTwo!!.setOnForfeitClickListener(this)
        controller = GameInfoController(this, Database.get(), Storage.teamId, Storage.userId)
        return fragmentView
    }

    override fun onDestroy() {
        super.onDestroy()
        controller?.unsubscribe()
        controller = null
    }

    override fun onPlayer1Info(user: User, ticket: Ticket) {
        activity.runOnUiThread { playerOne!!.update(user, ticket) }
    }

    override fun onPlayer2Info(user: User, ticket: Ticket) {
        activity.runOnUiThread { playerTwo!!.update(user, ticket) }
    }

    override fun getTitleResId(): Int {
        return R.string.app_name
    }

    override fun getAdvanceButtonTextResId(): Int {
        return R.string.next
    }

    override fun onPlayerForfeitClick(ticket: Ticket) {
        ticketController.ticket = ticket
        ticketController.transition(GAME_INFO, FORFEIT)
    }

    override fun onPlayerShareClick(ticket: Ticket) {
        ShareCompat.IntentBuilder
                .from(activity)
                .setText(ticket.code())
                .setType("text/plain")
                .setChooserTitle(R.string.share_code)
                .startChooser()
    }
}
